### ISSUE TITLE
[R] Reasearch embeddings

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#189]

**Comment by michalharakal** [comment]
```python
# word2vec_sgns.py
# A tiny, educational implementation of Word2Vec (Skip-gram with Negative Sampling) in PyTorch.
# Runs on CPU in a few minutes on a tiny toy corpus. For real corpora, increase epochs and batch size.

import math
import random
from collections import Counter, defaultdict

import torch
from torch import nn
from torch.utils.data import Dataset, DataLoader

# ---------------------------
# 1) Toy corpus & tokenization (replace with your own)
# ---------------------------
corpus = [
    "king queen man woman prince princess royal crown throne kingdom",
    "man walks to the store",
    "woman travels to the market",
    "king and queen rule the kingdom",
    "the prince wears a crown",
    "the princess sits on the throne",
    "royal family lives in the castle",
    "market has fresh food",
    "store sells many goods",
]

def tokenize(text):
    return text.lower().split()

tokens = [tokenize(line) for line in corpus]
all_tokens = [t for line in tokens for t in line]

# ---------------------------
# 2) Build vocabulary
# ---------------------------
min_count = 1
counts = Counter(all_tokens)
vocab = [w for w, c in counts.items() if c >= min_count]
vocab = sorted(vocab)  # just to have deterministic ids
stoi = {w: i for i, w in enumerate(vocab)}
itos = {i: w for w, i in stoi.items()}
freqs = torch.tensor([counts[itos[i]] for i in range(len(vocab))], dtype=torch.float)

# ---------------------------
# 3) Prepare positive (center, context) pairs
# ---------------------------
window_size = 2  # small for demo
pairs = []
for line in tokens:
    idxs = [stoi[w] for w in line if w in stoi]
    for i, center in enumerate(idxs):
        left = max(0, i - window_size)
        right = min(len(idxs), i + window_size + 1)
        for j in range(left, right):
            if j == i:
                continue
            pairs.append((center, idxs[j]))

# ---------------------------
# 4) Negative sampling distribution (unigram^0.75)
# ---------------------------
unigram = freqs.pow(0.75)
neg_dist = unigram / unigram.sum()

# Efficient alias table sampler (optional but fast); simple multinomial is fine for small vocabs
class NegativeSampler:
    def __init__(self, probs):
        self.probs = probs

    def sample(self, num_samples):
        # For small vocabs, multinomial is simplest
        return torch.multinomial(self.probs, num_samples=num_samples, replacement=True)

neg_sampler = NegativeSampler(neg_dist)

# ---------------------------
# 5) Dataset + collate: on-the-fly negative sampling
# ---------------------------
class SGNSDataset(Dataset):
    def __init__(self, pos_pairs):
        self.pos_pairs = pos_pairs

    def __len__(self):
        return len(self.pos_pairs)

    def __getitem__(self, idx):
        return self.pos_pairs[idx]

def collate_sgns(batch, num_negatives=5):
    centers = torch.tensor([c for c, _ in batch], dtype=torch.long)
    positives = torch.tensor([p for _, p in batch], dtype=torch.long)
    # negatives: for each (center, positive) we draw k negatives
    negatives = neg_sampler.sample(len(batch) * num_negatives).view(len(batch), num_negatives)
    return centers, positives, negatives

dataset = SGNSDataset(pairs)
loader = DataLoader(dataset, batch_size=128, shuffle=True, collate_fn=lambda b: collate_sgns(b, num_negatives=5))

# ---------------------------
# 6) Model: input & output embeddings + SGNS loss
# ---------------------------
class Word2VecSGNS(nn.Module):
    def __init__(self, vocab_size, embed_dim):
        super().__init__()
        self.in_embed = nn.Embedding(vocab_size, embed_dim)
        self.out_embed = nn.Embedding(vocab_size, embed_dim)
        self._init_emb()

    def _init_emb(self):
        # small uniform init like the original word2vec
        init_range = 0.5 / self.in_embed.embedding_dim
        nn.init.uniform_(self.in_embed.weight, -init_range, init_range)
        nn.init.zeros_(self.out_embed.weight)  # Mikolov often init'd output small/zeros

    def forward(self, centers, positives, negatives):
        """
        centers:  (B,)
        positives: (B,)
        negatives: (B, K)
        """
        v = self.in_embed(centers)            # (B, D)
        u_pos = self.out_embed(positives)     # (B, D)
        u_neg = self.out_embed(negatives)     # (B, K, D)

        # Positive score: log σ(v·u_pos)
        pos_score = torch.sum(v * u_pos, dim=1)                 # (B,)
        pos_loss = torch.log(torch.sigmoid(pos_score) + 1e-9)   # (B,)

        # Negative score: sum_k log σ(-v·u_neg_k)
        neg_score = torch.bmm(u_neg, v.unsqueeze(2)).squeeze(2) # (B, K)
        neg_loss = torch.sum(torch.log(torch.sigmoid(-neg_score) + 1e-9), dim=1)  # (B,)

        # Maximize pos + neg -> minimize negative of it
        loss = -(pos_loss + neg_loss)                           # (B,)
        return loss.mean()

    def get_input_weights(self):
        return self.in_embed.weight.detach()

# ---------------------------
# 7) Train
# ---------------------------
device = torch.device("cpu")
embed_dim = 50
model = Word2VecSGNS(len(vocab), embed_dim).to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

epochs = 10  # for demo; use 5-10+ on real corpora
for epoch in range(1, epochs + 1):
    total = 0.0
    for centers, positives, negatives in loader:
        centers, positives, negatives = centers.to(device), positives.to(device), negatives.to(device)
        loss = model(centers, positives, negatives)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        total += loss.item() * centers.size(0)
    print(f"Epoch {epoch}/{epochs} - loss: {total / len(dataset):.4f}")

# ---------------------------
# 8) Query: nearest neighbors by cosine similarity
# ---------------------------
@torch.no_grad()
def most_similar(query_word, topk=5):
    if query_word not in stoi:
        return []
    W = model.get_input_weights()  # (V, D)
    w = W[stoi[query_word]]        # (D,)
    sims = torch.mv(W, w) / (W.norm(dim=1) * (w.norm() + 1e-9))
    vals, idxs = torch.topk(sims, k=topk + 1)
    out = []
    for score, idx in zip(vals.tolist(), idxs.tolist()):
        word = itos[idx]
        if word == query_word:
            continue
        out.append((word, float(score)))
        if len(out) == topk:
            break
    return out

print("\nNearest neighbors:")
for q in ["king", "queen", "man", "woman", "market", "store"]:
    print(q, "->", most_similar(q, topk=5))
```

### how to use it

1. drop the file into a Python environment with `torch` installed.
2. run it (`python word2vec_sgns.py`).
3. replace the `corpus` list with your own sentences (or stream/clean a larger text), bump `epochs`, maybe raise `window_size`, `batch_size`, and `embed_dim`.

### notes & tweaks

* **Negative samples**: controlled by `num_negatives` in the collate function; common values are 5–20.
* **Speed**: for big corpora use a larger batch size, `nn.EmbeddingBag`, and consider subsampling frequent tokens (the classic `t = 1e-5` trick).
* **CBOW**: invert the direction—average context embeddings to predict the center word and reuse the same SGNS loss.
* **Saving vectors**: the learned word vectors are `model.in_embed.weight`. Save with `torch.save(model.state_dict(), "w2v.pth")`.

**Comment by aharakal** [comment]
#187 implemented. Embeddings implemented. Research resolved.
